Adding client builder for simple initialization of the Rust client

partylikeits1983 opened the pull request:
### Adding client builder for simple initialization of the Rust client

#### Overview

This PR introduces a new client builder to the Rust client, streamlining the initialization process and reduces the amount of boilerplate code when getting started with a Miden dApp. With the new client builder, setting up a Miden dApp in Rust becomes significantly easier. Personally, this was one of the features I really wanted to have as I was building the tutorial code examples.

By integrating this feature, we lower the barrier to entry and reduce the lines of code needed to deploy and interact with smart contracts and notes. For context, our [Miden dApp tutorials](https://0xpolygonmiden.github.io/miden-docs/miden-tutorials/rust-client/counter_contract_tutorial.html) currently showcase the older, more verbose approach where the `initialize_client()` function is used in each code example.

This improvement was discussed internally in [this issue](https://github.com/0xPolygonMiden/miden-devrel/issues/39).

#### What This Solves

- **Reduced Boilerplate:** Eliminates the need for the repetitive `initialize_client()` function in the tutorial docs.
- **Simpler Onboarding:** New users can now instantiate the client with minimal configuration.
- **Enhanced Flexibility:** The builder allows for easy configuration of RPC endpoints, timeouts, store paths, and debug modes.
- **Reduces Lines of Code:** The builder reduces the lines of code it takes to build a simple dApp on Miden using the rust client.

#### Usage

Initializing the client is now as simple as:

```rust
let mut client = Client::initialize()
    .with_rpc("https://rpc.testnet.miden.io:443")
    .in_debug_mode(true)
    .build()
    .await?;
```

#### Tests 

I added a simple test of instantiating the client using the new client builder. I also tested the new client builder in a [separate repository](https://github.com/partylikeits1983/miden-tx).

#### Future Work

There may be further improvements in the instantiation process. However, the primary goal here is to reduce the number of lines and the complexity required to get started with building on Miden.

---

Please let me know if I can make any improvements.

partylikeits1983 commented on crates/rust-client/src/tests.rs:
Is it ok to use `#[tokio::test]` here? And is it ok that this test requires an internet connection to ping the testnet?

igamigo reviewed (commented):
Thanks for the PR! I think a builder could be useful as the client's instantiation API becomes more complex. I had [made some comments a while ago about potentially introducing one](https://github.com/0xPolygonMiden/miden-client/issues/467#issuecomment-2505096125), but it's probably a very stale comment. 

In general we might not be able to provide "good" defaults (for the context of the tutorials, at least), because we need to be generic enough that the `Client` can be compiled for `no_std` environments, so not entirely sure if this would provide a ton of value for your use-case, but I still think it probably makes for a good addition.

partylikeits1983 commented on crates/rust-client/src/lib.rs:
```diff
@@ -271,13 +250,15 @@ impl<R: FeltRng> Client<R> {
         self.in_debug_mode
     }
 
-    /// Returns a reference to the client's random number generator. This can be used to generate
-    /// randomness for various purposes such as serial numbers, keys, etc.
```
Accidentally deleted it and then added back my own comment. Will change back to the original.

partylikeits1983 commented on crates/rust-client/src/lib.rs:
```diff
@@ -295,3 +276,12 @@ impl<R: FeltRng> Client<R> {
     }
 }
+
+// --------------------------------------------------------------------------------------------
+// BUILDER ENTRY POINT (OPTIONAL HELPER)
```
Fixed.

partylikeits1983 commented on crates/rust-client/src/lib.rs:
```diff
@@ -119,6 +98,9 @@ pub mod tests;
 
 mod errors;
 
+mod builder;
+pub use builder::ClientBuilder;
```
Fixed.

partylikeits1983 commented on crates/rust-client/src/builder.rs:
```diff
@@ -0,0 +1,103 @@
+    }
+
+    /// Sets the RPC endpoint via a URL.
+    pub fn with_rpc(mut self, url: &str) -> Self {
```
What about putting it behind a tonic feature flag? This would mean we could still compile for `wasm32`.

```rust
#[cfg(feature = "tonic")]
pub fn with_rpc(mut self, url: &str) -> Self {
    ...
}
```

igamigo replied:
I think we could add a `ClientBuilder` block within the `TonicRpcClient` that contains this method and we could avoid having the `#[cfg(feature="std")]` guard here.
Could look something like this:
```rust
// (on rpc/tonic_client/mod.rs):
impl ClientBuilder {
    #[must_use]
    pub fn with_tonic_rpc(mut self, url:&Endpoint, timeout: u64) -> Self {
        let tonic_client = TonicRpcClient::new(..,);
        self.with_rpc_client(Box::new(tonic_client))
    }
}
```

igamigo replied:
Maybe the same could be done for SQLite store. I think this might be a bit of an antipattern in Rust but it should work well here for making the builder available but also making it more ergonomic if certain features are turned on.

partylikeits1983 commented on crates/rust-client/src/builder.rs:
```diff
@@ -0,0 +1,103 @@
+        Self {
+            rpc_api: None,
+            timeout_ms: 10_000,
+            store_path: Some("store.sqlite3".into()),
```
Fixed.

partylikeits1983 commented:
1) Refactored the `src/builder.rs` to not use `std`. 
2) Added the builder test to the integration test directory.
3) Fixed small nits regarding comments.

partylikeits1983 commented:
Testing again locally with my [miden dApp testing repo](https://github.com/partylikeits1983/miden-tx), this PR makes client instantiation much cleaner! 

Now it is possible to do this:
```rust
let mut client = Client::initialize()
    .with_rpc("https://rpc.testnet.miden.io:443")
    .in_debug_mode(true)
    .build()
    .await?;
```

Instead of this: https://github.com/partylikeits1983/miden-tx/blob/ad0e378687ce950078b0bccd62ca030ac0ff09fb/src/bin/ephemeral_notes.rs#L26-L53

partylikeits1983 commented on crates/rust-client/src/builder.rs:
```diff
@@ -0,0 +1,103 @@
+use miden_objects::crypto::rand::RpoRandomCoin;
+use rand::Rng;
+
+pub struct ClientBuilder {
```
Added doc comments. Let me know if this is the correct format.

tomyrd commented on crates/rust-client/src/builder.rs:
```diff
@@ -0,0 +1,190 @@
+
+        // Create the endpoint using the parsed scheme, host, and port.
+        let endpoint = Endpoint::new(scheme.to_string(), host, port);
+        self.rpc_api = Some(Box::new(TonicRpcClient::new(endpoint, self.timeout_ms)));
+        self
```
nit
```suggestion
        self.with_rpc_api(Box::new(TonicRpcClient::new(endpoint, self.timeout_ms))));
```

partylikeits1983 replied:
> I think so too, this function could even receive a built Endpoint object. This way the user can use the known endpoints (like Endpoint::testnet()).

The only thing about doing this is that it might make initializing the client take one extra line of code in a Miden app. 

Would it mean doing this?
```rust
let endpoint = Endpoint::testnet();
let mut client = Client::initialize()
    .with_rpc(endpoint)
    .in_debug_mode(true)
    .build()
    .await?;
```

tomyrd replied:
You could create the endpoint inside the function call. Like this:
```rust
let mut client = Client::initialize()
    .with_rpc(Endpoint::testnet())
    .in_debug_mode(true)
    .build()
    .await?;
```

partylikeits1983 replied:
Ok this makes sense. Will change.

tomyrd commented on crates/rust-client/src/builder.rs:
```diff
@@ -0,0 +1,190 @@
+        self
+    }
+
+    /// Optionally set a custom timeout (in ms).
+    pub fn with_timeout(mut self, timeout_ms: u64) -> Self {
+        self.timeout_ms = timeout_ms;
+        self
+    }
```
We would need to specify that this function should be called before `with_rpc` (if not, the parameter would be ignored). I think it would be even better if we move the `timeout_ms` parameter to the `with_rpc` function. The parameter could be optional if we want to allow for a default timeout.

partylikeits1983 commented on crates/rust-client/src/builder.rs:
```diff
@@ -0,0 +1,190 @@
+    /// It parses the provided URL to extract the scheme, host, and port, and creates a `TonicRpcClient`.
+    #[cfg(feature = "tonic")]
+    pub fn with_rpc(mut self, url: &str) -> Self {
+        // Determine the scheme and strip it from the URL.
+        let (scheme, rest) = if let Some(stripped) = url.strip_prefix("https://") {
+            ("https", stripped)
+        } else if let Some(stripped) = url.strip_prefix("http://") {
+            ("http", stripped)
+        } else {
+            ("https", url)
+        };
+
+        // Attempt to find a colon indicating a port.
+        let (host, port) = if let Some(colon_index) = rest.find(':') {
+            // Split the host and port.
+            let host = &rest[..colon_index];
+            let port_str = &rest[colon_index + 1..];
+            // Try parsing the port. If it fails, use None.
+            let port = port_str.parse::<u16>().ok();
+            (host.to_string(), port)
+        } else {
+            // No colon found, so use the entire string as the host and no port.
+            (rest.to_string(), None)
+        };
+
+        // Create the endpoint using the parsed scheme, host, and port.
+        let endpoint = Endpoint::new(scheme.to_string(), host, port);
```
Yes it does! Will fix.

tomyrd commented on crates/rust-client/src/builder.rs:
```diff
@@ -0,0 +1,190 @@
+        };
+
+        // Create the authenticator.
+        let authenticator = StoreAuthenticator::new_with_rng(arc_store.clone(), rng.clone());
```
After merging #718, the authenticator now needs a `KeyStore` to be built, this keystore would need to be created externally and not by the builder (because the user of the client needs the keystore to add new keys). 

We can probably add a `with_keystore` function to the builder and use the provided keystore here. It should fail if no keystore is defined as there's no clear default implementation (one is std and the other no-std and they don't share parameters).

partylikeits1983 replied:
@tomyrd I can rebase this to the `next` branch if it makes things easier. Would be useful to have this feature on `0.7`, but if it's too complicated can rebase to `next`

tomyrd replied:
Thanks! But I also think it's best for 0.7, it won't be difficult to bring these changes to `next` once they get merged. I was just mentioning it so we don't forget.

partylikeits1983 replied:
Just saw this. @igamigo told me to rebase to `next`. Not an urgent change or anything. Rebasing to `next` means the PR can be ironed out a bit more.

partylikeits1983 commented:
Rebased this commit from `main` to `next`.

Also I made the entire client builder abstraction to only compile when `std` is available.

bobbinth commented on crates/rust-client/src/builder.rs:
```diff
@@ -0,0 +1,155 @@
+    }
+}
+
+impl ClientBuilder {
```
Should there also be a method for setting the keystore?

bobbinth reviewed (commented):
Looks great! Thank you for doing this. I left one small comment above.

Also, what needs to be done to make the CI green?

igamigo commented on crates/rust-client/src/lib.rs:
```diff
@@ -295,3 +273,12 @@ impl<R: FeltRng> Client<R> {
+    pub fn initialize() -> ClientBuilder {
+        ClientBuilder::new()
+    }
+}
```
I agree, is there any benefit to this helper?

partylikeits1983 replied:
I see now, definitely makes sense to remove it.

igamigo reviewed (commented):
Thanks! Leaving some more comments for now. As @bobbinth mentioned, the CI checks should be reviewed. Upon a quick look, errors seem quite simple. You can check them locally using the make targets (`make clippy`, `make clippy-wasm`, `make fix` for applying superficial fixes, etc.). Feel free to let me know if you have any questions about any of this

bobbinth reviewed (approved):
Looks good! Thank you! It seems like the integration tests are not passing - but maybe that's something can be fixed rebasing from `next`?

tomyrd commented on crates/rust-client/src/builder.rs:
```diff
@@ -0,0 +1,187 @@
+        // Use the provided keystore, or default to a FilesystemKeyStore with temp_dir.
+        let keystore = match self.keystore {
+            Some(ks) => ks,
+            None => FilesystemKeyStore::new(temp_dir())
+                .map_err(|err| ClientError::ClientInitializationError(err.to_string()))?,
```
Should we fail if the keystore is not set? If not, I wouldn't use the `temp_dir` to store the keys, it might get cleared up.

partylikeits1983 replied:
Ah yes that is true. What would be a good reasonable default? I think I was following the cli tests as a guide when writing this, that's why I use `temp_dir()`.

tomyrd replied:
In the CLI we create a "keystore" dir on the current directory and use that. 

But in this case, I feel like we should have it so that the user always provides the keystore themselves. My reasoning is that if we create the keystore in the builder, then the user won't have a reference to it. This means that when adding new accounts, the user won't be able to add the authentication keys to the keystore.

partylikeits1983 replied:
Would this make sense as a standard use example? This creates a `/keystore` directory.

```rust
let keystore = FilesystemKeyStore::new("keystore".into()).unwrap();
let mut client = ClientBuilder::new()
    .with_rpc(Endpoint::try_from("https://rpc.testnet.miden.io:443").unwrap())
    .with_timeout(10_000)
    .with_keystore(keystore)
    .with_sqlite_store("store.sqlite3")
    .in_debug_mode(true)
    .build()
    .await?;
```

partylikeits1983 replied:
This way the user has reference to the keystore.

tomyrd replied:
If we make it so that the `with_keystore()` method takes a string, then we would have to change it so that the builder also returns the keystore it created internally, something like:
```
let (mut client, mut keystore) = ClientBuilder::new()
    .with_rpc(Endpoint::try_from("https://rpc.testnet.miden.io:443").unwrap())
    .with_timeout(10_000)
    .with_keystore("keystore")
    .with_sqlite_store("store.sqlite3")
    .in_debug_mode(true)
    .build()
    .await?;
```
I'm open to either one but I think I prefer the first one.

bobbinth replied:
I think one of the goals here is to make assumptions about a reasonable set of defaults while also keeping some flexibility. So, I'd probably go with more specialized methods that make ergonomics a bit better. For example:
```rust
let mut client = ClientBuilder::new()
    .with_tonic_rpc("https://rpc.testnet.miden.io:443")
    .with_timeout(10_000)
    .with_filesystem_keystore("keystore")
    .with_sqlite_store("store.sqlite3")
    .in_debug_mode(true)
    .build()
    .await?;
```
Ff the user wanted to customize things beyond these defaults, they could also do something like:
```rust
let mut client = ClientBuilder::new()
    .with_rpc(my_rpc)
    .with_keystore(my_keystore)
    .with_store(my_store)
    .in_debug_mode(true)
    .build()
    .await?;
```
The only thing which I think doesn't fit this pattern would be the random coin. But maybe we keep it as is for now and make it more flexible in another PR.

tomyrd reviewed (approved):
LGTM! Left a small comment to be addressed before merging.

SantiagoPittella reviewed (commented):
Overall looks good, I think that after addressing https://github.com/0xPolygonMiden/miden-client/pull/741/files#r1962738199 and https://github.com/0xPolygonMiden/miden-client/pull/741/files#r1960564185 this can be ready to merge.

bobbinth commented on crates/rust-client/src/builder.rs:
```diff
@@ -0,0 +1,103 @@
+    }
+
+    /// Optionally set a custom store path.
+    pub fn with_store_path(mut self, path: &str) -> Self {
```
I think the more generic method is `with_store()` right? I think one of the intentions of this method was it to make it configuring sqlite store as easy as possible. So, maybe the way to go there is to rename this into something like `with_sqlite_store()` so that it is clear that this method is intended to work specifically with sqlite stores.

igamigo replied:
Ah yes, true. Agree with the rename. I'd also move it to the SQLite mod but maybe we could refactor it separately

partylikeits1983 commented on crates/rust-client/src/builder.rs:
```diff
@@ -0,0 +1,155 @@
+    /// An optional store provided by the user.
+    store: Option<Arc<dyn Store>>,
+    /// An optional RNG provided by the user.
+    rng: Option<RpoRandomCoin>,
```
Would this mean that when initializing the client the user will have to specify type annotations for the client?

I implemented a basic version making `RpoRandomCoin` generic, but it requires that I specify type annotations when initializing the client. 

This is how it looks:
```rust
let mut client = ClientBuilder::<RpoRandomCoin>::new()
    .with_rpc(Endpoint::try_from("https://rpc.testnet.miden.io:443").unwrap())
    .with_timeout(10_000)
    .with_sqlite_store("store.sqlite3")
    .in_debug_mode(true)
    .build()
    .await?;
```

Maybe there is a way to make `RpoRandomCoin` generic and not have to specify the type annotation?

partylikeits1983 commented:
Addressed several comments from above.

1) `keystore` must be specified during build phase.
2) `ClientBuilder` now has 2 methods for setting the `keystore`: `.with_keystore()` and `.with_filesystem_keystore()`

The following is possible:
```rust
let mut client = ClientBuilder::new()
    .with_tonic_rpc("https://rpc.testnet.miden.io:443")
    .with_timeout(10_000)
    .with_filesystem_keystore("keystore")
    .with_sqlite_store("store.sqlite3")
    .in_debug_mode(true)
    .build()
    .await?;
    ```
The `.with_keystore()` method allows the user to directly set the keystore.

partylikeits1983 commented:
Force pushed to keep everything in a single commit & to fix CI/CD issues.

partylikeits1983 commented:
@igamigo Regarding this: https://github.com/0xPolygonMiden/miden-client/pull/741/files#r1962738199

Would it be possible to address this in a separate PR? I tried implementing it but I believe it could get a bit complicated. Also not 100% sure how to make `RpoRandomCoin` generic without having to specify the type in the `ClientBuilder`, which means my initial implementation made things look like this: 
```rust
let mut client = ClientBuilder::<RpoRandomCoin>::new();
```

bobbinth commented on crates/rust-client/src/builder.rs:
```diff
@@ -0,0 +1,224 @@
+/// Without this enum, you would be forced to perform the initialization immediately, which would
+/// require unwrapping (and potentially panicking) if initialization fails. By deferring it, we
+/// allow error handling during `build()`.
+pub enum KeystoreConfig {
+    Path(String),
+    Instance(FilesystemKeyStore),
+}
```
Does this need to be public? Overall, I think this is fine for now but we'd probably want to refactor how we handle keystores a bit the future. The main idea would be so that we could provide the keystore dynamically (similar to how we do it with the `Store`). But how to do it will depend on what we do in https://github.com/0xPolygonMiden/miden-client/issues/742.

partylikeits1983 replied:
Fixed.

bobbinth commented on crates/rust-client/src/builder.rs:
```diff
@@ -0,0 +1,224 @@
+    /// used during the build process, together with the timeout. This means that any call to
+    /// `with_timeout` will be effective as long as it happens before `build()`.
+    #[must_use]
+    pub fn with_rpc(mut self, endpoint: Endpoint) -> Self {
+        self.rpc_endpoint = Some(endpoint);
+        self
+    }
+
+    /// Sets a custom RPC client directly.
+    ///
+    /// This method overrides any previously set RPC endpoint.
+    #[must_use]
+    pub fn with_rpc_client(mut self, client: Box<dyn NodeRpcClient + Send>) -> Self {
```
nit: I would maybe rename the current `with_rpc_client()` into just `with_rpc()`, and the current `with_rpc()` into `with_tonic_rpc()`.

Also, `with_tonic_rpc()` could probably take a string rather than `Endpoint` as an argument - or is there a reason not to do this?

partylikeits1983 replied:
> Also, with_tonic_rpc() could probably take a string rather than Endpoint as an argument - or is there a reason not to do this?

The reason why the `with_rpc()` function takes an inputs was because I was following @tomyrd suggestion here:
> I think so too, this function could even receive a built Endpoint object. This way the user can use the known endpoints (like Endpoint::testnet())
https://github.com/0xPolygonMiden/miden-client/pull/741#discussion_r1961857181

I do like the idea of just passing in the RPC url string as it is cleaner, but maybe it is useful to be able to initialize the endpoing like: `Endpoint::testnet()`

bobbinth replied:
Yep, makes sense. I'm fine with either approach.

bobbinth replied:
btw, another option is to use something like `AsRef<str>` or maybe `Into<String>` - e.g.:
```rust
pub fn with_tonic_rpc(mut self, url: impl Into<String>) -> Self {
    ...
}
```
Then, if we have `From<Endpoint> for String` implemented, we'd be able to pass it into the method as well. We would need to change the internal field type of `rpc_endpoint` to `String`, but I don't think this should be a problem.

bobbinth reviewed (approved):
Looks good! Thank you! I left a coupe of optional comments inline.

Beyond these, I think we can make a couple of improvements in the future:
- Make handling of `KeyStore` more generic (so that we could use non-`FilesystemKeyStore`s as well).
- Make handling of `Rng` more generic (so that we could use non-`RpoRandomCoin` as well).

But I would do it in a separate PR - so, let's create an issue for this.

partylikeits1983 commented:
1) Changed the name of the function `with_rpc_client()` to `with_tonic_client()` as suggested by @bobbinth  
2) Added a test for using the `ClientBuilder` with `with_tonic_client()`
3) Rebased to latest `next` branch updates.

partylikeits1983 commented:
Opened an issue for a separate PR to improve the `ClientBuilder` implementation. https://github.com/0xPolygonMiden/miden-client/issues/759

bobbinth commented on crates/rust-client/src/builder.rs:
```diff
@@ -0,0 +1,224 @@
+    /// used during the build process, together with the timeout. This means that any call to
+    /// `with_timeout` will be effective as long as it happens before `build()`.
+    #[must_use]
+    pub fn with_rpc(mut self, endpoint: Endpoint) -> Self {
+        self.rpc_endpoint = Some(endpoint);
+        self
+    }
+
+    /// Sets a custom RPC client directly.
+    ///
+    /// This method overrides any previously set RPC endpoint.
+    #[must_use]
+    pub fn with_tonic_rpc(mut self, client: Box<dyn NodeRpcClient + Send>) -> Self {
```
I think the naming here is reversed. It should be:
```rust
pub fn with_rpc(mut self, client: Box<dyn NodeRpcClient + Send>) -> Self {
    ...
}

pub fn with_tonic_rpc(mut self, endpoint: Endpoint) -> Self {
    ...
}
```
The idea is that `with_tonic_rpc()` is more specialized - i.e., we know that we'll be using the `TonicRpcClient`.

partylikeits1983 replied:
That makes sense. I changed the naming of the functions.

igamigo commented:
Unused dependencies check was fixed on #761 so we can merge this safely
